File: back/src/main/java/fr/entity/Prestation.java
package fr.entity;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Getter
@Setter

@Table(name="`prestation`")
public class Prestation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "user_id")
    @JsonIgnore
    private User user;

    private String title;
    private Long duration;
    private Integer addPoint;
    private Timestamp dateStart;
    private Timestamp dateEnd;
    private String state;
    private Integer maxUser;
    private Integer placeAvailable;
    private String videoLink;

    @Column(columnDefinition = "TEXT")
    private String description;
    
    private String littleDescription;

    @Column(columnDefinition = "TEXT")
    private String practicalInformation;

    private String language;

    @Column(columnDefinition = "TEXT")
    private String personalInfos;

    @OneToMany(mappedBy = "prestation", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<Image> images = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "type_id")
    @JsonIgnore
    private Type type;

    @OneToMany(mappedBy = "prestation", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<Registration> registrations = new ArrayList<>();
    
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "location_id")
    @JsonIgnore
    private Location location;


    public Prestation() {
        this.placeAvailable = 0;
    }



    public Prestation(Integer id, String title, Long duration, Integer addPoint, Timestamp dateStart, Timestamp dateEnd, String state, Integer maxUser, Integer placeAvailable, String description, String littleDescription, String practicalInformation, String language, String personalInfos, List<Image> images, Type type, List<Registration> registrations, Location location) {
        this.id = id;
        this.title = title;
        this.duration = duration;
        this.addPoint = addPoint;
        this.dateStart = dateStart;
        this.dateEnd = dateEnd;
        this.state = state;
        this.maxUser = maxUser;
        this.placeAvailable = placeAvailable;
        this.description = description;
        this.littleDescription = littleDescription;
        this.practicalInformation = practicalInformation;
        this.language = language;
        this.personalInfos = personalInfos;
        this.images = images;
        this.type = type;
        this.registrations = registrations;
        this.location = location;
    }




    public void bookedPlace() {
        this.setPlaceAvailable(this.getPlaceAvailable()-1); 
    }

    public void setMaxUser(Integer maxUser) {
        this.maxUser = maxUser;
        calculatePlaceavailable();
    }

    private void calculatePlaceavailable() {
        if (this.getRegistrations() != null && this.getMaxUser() > 0) {
            this.setPlaceAvailable(this.getMaxUser() - this.getRegistrations().size());
        }
    }
    
     
}

File: front/src/app/shared/model/prestation.ts
import { Type } from "./type";
import { Registration } from "./registration";
import { Location } from "./location";
import { Image } from "./image";
import { User } from "./user";
import { Category } from "./category";

export class Prestation {

    constructor(
        public id? : number,
        public user? : User,
        public title : string ="",
        public duration? : number,
        public addPoint? : number,
        public dateStart? : number,
        public dateEnd? : number,
        public state? : string,
        public description? : string,
        public maxUser? : number,
        public images : Image[] = [],
        public type?: Type,
        public category?: Category,
        public location?: Location,
        public placeAvailable?: number,
        public littleDescription?: string,
        public videoLink?: string,
        public practicalInformation?: string,
        public language?: string,
        public personalInfos?: string,
        public registrations?: Registration[],
    ){}

  get getDuration(): string {
    if (this.duration != undefined) {
      const durautionHour = this.duration / 1000 / 60 / 60;
      return durautionHour.toString();
    }
    return '0';
  }

    get getDateStartString(): string {
        if (typeof this.dateStart === 'number') {
            const dateStart = new Date(this.dateStart);
            const day = dateStart.getDate().toString().padStart(2, '0');
            const month = (dateStart.getMonth() + 1).toString().padStart(2, '0');
            const year = dateStart.getFullYear().toString();
            const formattedDate = `${day}/${month}/${year}`;
            return formattedDate;
        }
        return 'dd/mm/yyyy'; 
    }
      
    get getTimeStartString(): string {
        if (typeof this.dateStart === 'number') {
            const date = new Date(this.dateStart);
            const hours = date.getHours().toString().padStart(2, '0');
            const minutes = date.getMinutes().toString().padStart(2, '0');
            return `${hours}:${minutes}`;
        }
        return "hh:mm";
          
    }

    get getDateTimeStartString(): string {
        if (typeof this.dateStart === 'number') {
            const dateTime = new Date(this.dateStart);    
            const day = dateTime.getDate().toString().padStart(2, '0');
            const month = (dateTime.getMonth() + 1).toString().padStart(2, '0');
            const year = dateTime.getFullYear().toString();
            const hours = dateTime.getHours().toString().padStart(2, '0');
            const minutes = dateTime.getMinutes().toString().padStart(2, '0');
      
        return `${day} ${month} ${year}, ${hours}:${minutes}`;
      }
      return "dd MM yyyy, HH:mm";
      
    }

}
